include/dr/detail/communicator.hpp

```diff
@@ -124,12 +124,12 @@ class communicator {
                  rng::data(recvr), count * sizeof(T), MPI_BYTE, mpi_comm_);
   }
 
-  template <rng::contiguous_range R>
-  void alltoallv(const R &sendbuf, const std::vector<std::size_t> &sendcnt,
-                 const std::vector<std::size_t> &senddsp, R &recvbuf,
+  template <rng::contiguous_range R1, rng::contiguous_range R2>
+  void alltoallv(const R1 &sendbuf, const std::vector<std::size_t> &sendcnt,
+                 const std::vector<std::size_t> &senddsp, R2 &recvbuf,
                  const std::vector<std::size_t> &recvcnt,
                  const std::vector<std::size_t> &recvdsp) {
-    using valT = typename R::value_type;
+    using valT = typename R2::value_type;
```

Comment:
Containers can be of different types, but the elements stored must be of the same type. However, there is no way to check this because not all containers implement value_type (ranges::subrange<int *> in our case).

Reply:
At least [decltype](https://en.cppreference.com/w/cpp/language/decltype) could be used to implement the check. But there is a nicer way: because not all containers implement value_type there is  [rng::range_value_t](https://en.cppreference.com/w/cpp/ranges/iterator_t) which should always work.